teamcity: failed tests on master: testrace/TestRangeTransferLease, testrace/TestRangeTransferLease/DrainTransferWithExtension

The following tests appear to have failed:

[#311912](https://teamcity.cockroachdb.com/viewLog.html?buildId=311912):

```
--- FAIL: testrace/TestRangeTransferLease (45.700s)

------- Stdout: -------
I170801 19:29:36.369613 61585 storage/store.go:1271  [n1,s1]: failed initial metrics computation: [n1,s1]: system config not yet available
I170801 19:29:36.379535 61585 gossip/gossip.go:295  [n1] NodeDescriptor set to node_id:1 address:<network_field:"tcp" address_field:"127.0.0.1:34551" > attrs:<> locality:<> 
W170801 19:29:36.404274 61585 gossip/gossip.go:1204  [n2] no incoming or outgoing connections
I170801 19:29:36.407075 61334 gossip/client.go:129  [n2] started gossip client to 127.0.0.1:34551
I170801 19:29:36.428942 61585 gossip/gossip.go:295  [n2] NodeDescriptor set to node_id:2 address:<network_field:"tcp" address_field:"127.0.0.1:48662" > attrs:<> locality:<> 
I170801 19:29:36.453593 61585 storage/client_test.go:411  gossip network initialized
I170801 19:29:36.459397 61338 storage/replica_raftstorage.go:509  [s1,r1/1:/M{in-ax}] generated preemptive snapshot d63ac252 at index 15
I170801 19:29:36.463200 61338 storage/store.go:3486  [s1,r1/1:/M{in-ax}] streamed snapshot to (n2,s2):?: kv pairs: 33, log entries: 5, rate-limit: 2.0 MiB/sec, 3ms
I170801 19:29:36.464693 61785 storage/replica_raftstorage.go:718  [s2,r1/?:{-}] applying preemptive snapshot at index 15 (id=d63ac252, encoded size=6302, 1 rocksdb batches, 5 log entries)
I170801 19:29:36.466473 61785 storage/replica_raftstorage.go:726  [s2,r1/?:/M{in-ax}] applied preemptive snapshot in 2ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I170801 19:29:36.470063 61338 storage/replica_command.go:3600  [s1,r1/1:/M{in-ax}] change replicas (ADD_REPLICA (n2,s2):2): read existing descriptor r1:/M{in-ax} [(n1,s1):1, next=2]
I170801 19:29:36.476300 61660 storage/replica.go:2937  [s1,r1/1:/M{in-ax}] proposing ADD_REPLICA((n2,s2):2): updated=[(n1,s1):1 (n2,s2):2] next=3
I170801 19:29:36.677519 61923 storage/raft_transport.go:451  raft transport stream to node 1 established
--- FAIL: testrace/TestRangeTransferLease/DrainTransferWithExtension (45.130s)
<autogenerated>:14: storage/client_replica_test.go:542, condition failed to evaluate within 45s: [NotLeaseHolderError] r1: replica (n1,s1):1 not lease holder; current lease is repl=(n2,s2):2 start=0.900000123,159 exp=2.700000123,4 pro=1.800000123,5
------- Stdout: -------
I170801 19:30:01.487692 61826 gossip/gossip.go:1218  [n2] node has connected to cluster via gossip
I170801 19:30:09.870273 61699 gossip/gossip.go:1218  [n1] node has connected to cluster via gossip
I170801 19:30:22.033237 61689 vendor/google.golang.org/grpc/grpclog/grpclog.go:75  transport: http2Server.HandleStreams failed to read frame: read tcp 127.0.0.1:34551->127.0.0.1:36934: use of closed network connection
W170801 19:30:22.033579 61923 storage/raft_transport.go:457  raft transport stream to node 1 failed: rpc error: code = Unavailable desc = transport is closing

```
Please assign, take a look and update the issue accordingly.

--- a/pkg/base/config.go
+++ b/pkg/base/config.go
@@ -55,9 +55,9 @@ const (
 	// defaultRaftTickInterval is the default resolution of the Raft timer.
 	defaultRaftTickInterval = 200 * time.Millisecond
 
-	// rangeLeaseRaftElectionTimeoutMultiplier specifies what multiple the
+	// defaultRangeLeaseRaftElectionTimeoutMultiplier specifies what multiple the
 	// leader lease active duration should be of the raft election timeout.
-	rangeLeaseRaftElectionTimeoutMultiplier = 3
+	defaultRangeLeaseRaftElectionTimeoutMultiplier = 3
 
 	// rangeLeaseRenewalFraction specifies what fraction the range lease
 	// renewal duration should be of the range lease active time. For example,
@@ -320,6 +320,10 @@ type RaftConfig struct {
 	// previous election expires. This value is inherited by individual stores
 	// unless overridden.
 	RaftElectionTimeoutTicks int
+
+	// RangeLeaseRaftElectionTimeoutMultiplier specifies what multiple the leader
+	// lease active duration should be of the raft election timeout.
+	RangeLeaseRaftElectionTimeoutMultiplier float64
 }
 
 // SetDefaults initializes unset fields.
@@ -329,6 +333,9 @@ func (cfg *RaftConfig) SetDefaults() {
 	}
 	if cfg.RaftElectionTimeoutTicks == 0 {
 		cfg.RaftElectionTimeoutTicks = defaultRaftElectionTimeoutTicks
+	}
+	if cfg.RangeLeaseRaftElectionTimeoutMultiplier == 0 {
+		cfg.RangeLeaseRaftElectionTimeoutMultiplier = defaultRangeLeaseRaftElectionTimeoutMultiplier
 	}
 }
 
@@ -341,7 +348,8 @@ func (cfg RaftConfig) RaftElectionTimeout() time.Duration {
 // RangeLeaseDurations computes durations for range lease expiration and
 // renewal based on a default multiple of Raft election timeout.
 func (cfg RaftConfig) RangeLeaseDurations() (rangeLeaseActive, rangeLeaseRenewal time.Duration) {
-	rangeLeaseActive = time.Duration(rangeLeaseRaftElectionTimeoutMultiplier * float64(cfg.RaftElectionTimeout()))
+	rangeLeaseActive = time.Duration(cfg.RangeLeaseRaftElectionTimeoutMultiplier *
+		float64(cfg.RaftElectionTimeout()))
 	rangeLeaseRenewal = time.Duration(float64(rangeLeaseActive) * rangeLeaseRenewalFraction)
 	return
 }

--- a/pkg/storage/client_replica_test.go
+++ b/pkg/storage/client_replica_test.go
@@ -466,6 +466,10 @@ func TestRangeLookupUseReverse(t *testing.T) {
 func TestRangeTransferLease(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	cfg := storage.TestStoreConfig(nil)
+	// Ensure the node liveness duration isn't too short. By default it is 900ms
+	// for TestStoreConfig().
+	cfg.RangeLeaseRaftElectionTimeoutMultiplier =
+		float64((9 * time.Second) / cfg.RaftElectionTimeout())
 	var filterMu syncutil.Mutex
 	var filter func(filterArgs storagebase.FilterArgs) *roachpb.Error
 	cfg.TestingKnobs.TestingEvalFilter =
@@ -578,7 +582,18 @@ func TestRangeTransferLease(t *testing.T) {
 	forceLeaseExtension := func(sender *storage.Stores, lease roachpb.Lease) error {
 		shouldRenewTS := lease.Expiration.Add(-1, 0)
 		mtc.manualClock.Set(shouldRenewTS.WallTime + 1)
-		return sendRead(sender).GoError()
+		err := sendRead(sender).GoError()
+		if err != nil {
+			// We can sometimes receive an error from our renewal attempt because the
+			// lease transfer ends up causing the renewal to re-propose and second
+			// attempt fails because it's already been renewed. This used to work
+			// before we compared the proposer's lease with the actual lease because
+			// the renewed lease still encompassed the previous request.
+			if _, ok := err.(*roachpb.NotLeaseHolderError); ok {
+				err = nil
+			}
+		}
+		return err
 	}
 	t.Run("Transfer", func(t *testing.T) {
 		origLease, _ := replica0.GetLease()
@@ -651,7 +666,12 @@ func TestRangeTransferLease(t *testing.T) {
 		transferErrCh := make(chan error)
 		go func() {
 			// Transfer back from replica1 to replica0.
-			transferErrCh <- replica1.AdminTransferLease(context.Background(), replica0Desc.StoreID)
+			err := replica1.AdminTransferLease(context.Background(), replica0Desc.StoreID)
+			// Ignore not leaseholder errors which can arise due to re-proposals.
+			if _, ok := err.(*roachpb.NotLeaseHolderError); ok {
+				err = nil
+			}
+			transferErrCh <- err
 		}()
 		// Wait for the transfer to be blocked by the extension.
 		<-transferBlocked
@@ -660,16 +680,8 @@ func TestRangeTransferLease(t *testing.T) {
 		checkHasLease(t, mtc.senders[0])
 		setFilter(false, nil)
 
-		// We can sometimes receive an error from our renewal attempt
-		// because the lease transfer ends up causing the renewal to
-		// re-propose and second attempt fails because it's already been
-		// renewed. This used to work before we compared the proposer's lease
-		// with the actual lease because the renewed lease still encompassed the
-		// previous request.
 		if err := <-renewalErrCh; err != nil {
-			if _, ok := err.(*roachpb.NotLeaseHolderError); !ok {
-				t.Errorf("expected not lease holder error due to re-proposal; got %s", err)
-			}
+			t.Errorf("unexpected error from lease renewal: %s", err)
 		}
 		if err := <-transferErrCh; err != nil {
 			t.Errorf("unexpected error from lease transfer: %s", err)
@@ -717,7 +729,7 @@ func TestRangeTransferLease(t *testing.T) {
 		if !ok {
 			t.Fatalf("expected %T, got %s", &roachpb.NotLeaseHolderError{}, pErr)
 		}
-		if *(nlhe.LeaseHolder) != replica1Desc {
+		if nlhe.LeaseHolder == nil || *nlhe.LeaseHolder != replica1Desc {
 			t.Fatalf("expected lease holder %+v, got %+v",
 				replica1Desc, nlhe.LeaseHolder)
 		}
@@ -754,16 +766,8 @@ func TestRangeTransferLease(t *testing.T) {
 		checkHasLease(t, mtc.senders[0])
 		setFilter(false, nil)
 
-		// We can sometimes receive an error from our renewal attempt
-		// because the lease transfer ends up causing the renewal to
-		// re-propose and second attempt fails because it's already been
-		// renewed. This used to work before we compared the proposer's lease
-		// with the actual lease because the renewed lease still encompassed the
-		// previous request.
 		if err := <-renewalErrCh; err != nil {
-			if _, ok := err.(*roachpb.NotLeaseHolderError); !ok {
-				t.Errorf("expected not lease holder error due to re-proposal; got %s", err)
-			}
+			t.Errorf("unexpected error from lease renewal: %s", err)
 		}
 	})
 }

--- a/pkg/storage/replica_test.go
+++ b/pkg/storage/replica_test.go
@@ -765,9 +765,10 @@ func TestReplicaLease(t *testing.T) {
 
 	// Test that leases with invalid times are rejected.
 	// Start leases at a point that avoids overlapping with the existing lease.
-	one := hlc.Timestamp{WallTime: time.Second.Nanoseconds(), Logical: 0}
+	leaseDuration := tc.store.cfg.RangeLeaseActiveDuration()
+	start := hlc.Timestamp{WallTime: (time.Second + leaseDuration).Nanoseconds(), Logical: 0}
 	for _, lease := range []roachpb.Lease{
-		{Start: one, Expiration: hlc.Timestamp{}},
+		{Start: start, Expiration: hlc.Timestamp{}},
 	} {
 		if _, err := evalRequestLease(context.Background(), tc.store.Engine(),
 			CommandArgs{

--- a/pkg/storage/store.go
+++ b/pkg/storage/store.go
@@ -129,6 +129,9 @@ func TestStoreConfig(clock *hlc.Clock) StoreConfig {
 		HistogramWindowInterval:     metric.TestSampleInterval,
 		EnableEpochRangeLeases:      true,
 	}
+
+	// Use shorter Raft tick settings in order to minimize start up and failover
+	// time in tests.
 	sc.RaftElectionTimeoutTicks = 3
 	sc.RaftTickInterval = 100 * time.Millisecond
 	sc.SetDefaults()

